go/merkle/types.go

```diff
@@ -0,0 +1,173 @@
+package merkle
+
+import (
+	"crypto/sha512"
+
+	"github.com/keybase/client/go/msgpack"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/client/go/sig3"
+	merkletree "github.com/keybase/go-merkle-tree"
+	"github.com/pkg/errors"
+)
+
+type TreeSeqno int64
+
+type EncodingType byte
+
+const (
+	EncodingTypeBlindedSHA512_256v1 EncodingType = 1 // p = HMAC-SHA512-256; (k, v) -> (k, p(p(k, s), v)) where s is a secret unique per Merkle seqno
+)
+
+const CurrentEncodingType = EncodingTypeBlindedSHA512_256v1
+
+func GetTreeConfig(encodingType EncodingType) (merkletree.Config, error) {
+	switch encodingType {
+	case EncodingTypeBlindedSHA512_256v1:
+		return merkletree.NewConfig(SHA512_256Hasher{}, 32, 64, EncodedLeaf{}), nil
+	}
+	return merkletree.Config{}, errors.Errorf("unknown encoding type %q", encodingType)
+}
+
+type EncodedLeaf []byte
+
+var _ merkletree.ValueConstructor = (*EncodedLeaf)(nil)
+
+func (l EncodedLeaf) Construct() interface{} {
+	return &[]byte{}
+}
+
+type LeafType uint16
+
+const (
+	LeafTypeChain17v1 = 1
+)
+
+type LeafBytes []byte
+type LeafContainer struct {
+	_struct   bool      `codec:",toarray"`
+	LeafType  LeafType  // specifies structure of leafBytes
+	LeafBytes LeafBytes // msgpack deserialization implements Leaf
+}
+
+func NewLeafContainer(leafType LeafType, leafBytes LeafBytes) LeafContainer {
+	return LeafContainer{LeafType: leafType, LeafBytes: leafBytes}
+}
+
+func (c LeafContainer) Serialize() ([]byte, error) {
+	return msgpack.EncodeCanonical(c)
+}
+
+type ID []byte
+
+type Leaf interface {
+	Serialize() ([]byte, error)
+	Type() LeafType
+	ID() ID
+	GetSeqno() keybase1.Seqno
+}
+
+type SigID []byte
+type Chain17v1Leaf struct {
+	_struct bool `codec:",toarray"`
+
+	// do not encode teamID; it's redundant in the tree
+	TeamID sig3.TeamID `codec:"-"`
+
+	SigID  SigID
```

Comment:
Do we really need the sigID I wonder? We never in practice use them, only the `LinkID`s

Reply:
I don't think it's needed but it also doesn't hurt much to leave in, since it isn't in the tree and generated on demand

Reply:
Why doesn't this have `ChainType` too?  Where is that data going to live, assuming we have multiple chains for each team?

Also, it would be great if we can use the same code / representation as in `hidden/ratchet.go` but is that not possible?

Reply:
ChainType is in the type of the leaf. What's actually in the tree is this

```
type LeafContainer struct {
    _struct   bool      `codec:",toarray"`
    LeafType  LeafType  // specifies structure of leafBytes
    LeafBytes LeafBytes // msgpack deserialization implements Leaf
}
```
```
type LeafType uint16

const (
    LeafTypeChain17v1 = 1
)
```

>Also, it would be great if we can use the same code / representation as in hidden/ratchet.go but is that not possible?
Not sure exactly what you mean, do you want to use sig3.Tail here?

Reply:
Yeah, though the fields are different in a sig3.Tail. It's too bad that off the bat, we have two totally different types of ratchets. The fields in the `sig3.Tail` are:

```type Tail struct {
	Hash      LinkID    `codec:"h"` // hash of the outer link
	Seqno     Seqno     `codec:"s"`
	ChainType ChainType `codec:"t"`
}```